src/main/scala/io/iohk/ethereum/consensus/blocks/BlockGeneratorImpl.scala

```diff
@@ -0,0 +1,174 @@
+package io.iohk.ethereum.consensus.blocks
+
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.util.ByteString
+import io.iohk.ethereum.consensus.ConsensusConfig
+import io.iohk.ethereum.consensus.ethash.blocks.Ommers
+import io.iohk.ethereum.consensus.validators.std.MptListValidator.intByteArraySerializable
+import io.iohk.ethereum.crypto.kec256
+import io.iohk.ethereum.db.dataSource.EphemDataSource
+import io.iohk.ethereum.db.storage.{ArchiveNodeStorage, NodeStorage}
+import io.iohk.ethereum.domain._
+import io.iohk.ethereum.ledger.Ledger.{BlockPreparationResult, BlockResult}
+import io.iohk.ethereum.ledger.{BlockPreparator, BloomFilter}
+import io.iohk.ethereum.mpt.{ByteArraySerializable, MerklePatriciaTrie}
+import io.iohk.ethereum.network.p2p.messages.PV62.BlockBody
+import io.iohk.ethereum.network.p2p.messages.PV62.BlockHeaderImplicits._
+import io.iohk.ethereum.utils.BlockchainConfig
+import io.iohk.ethereum.utils.ByteUtils.or
+
+import scala.collection.immutable
+
+/**
+ * This is a skeleton for a generic [[io.iohk.ethereum.consensus.blocks.BlockGenerator BlockGenerator]].
+ *
+ * @param blockchain
+ * @param blockchainConfig
+ * @param _blockTimestampProvider
+ */
+abstract class BlockGeneratorImpl(
+  blockchain: Blockchain,
+  blockchainConfig: BlockchainConfig,
+  consensusConfig: ConsensusConfig,
+  blockPreparator: BlockPreparator,
+  _blockTimestampProvider: BlockTimestampProvider = DefaultBlockTimestampProvider
+) extends TestBlockGenerator {
+
+  protected val headerExtraData = consensusConfig.headerExtraData
+
+  protected val blockCacheSize = consensusConfig.blockCacheSize
+
+  protected val difficulty = new DifficultyCalculator(blockchainConfig)
+
+  protected val cache: AtomicReference[List[PendingBlockAndState]] = new AtomicReference(Nil)
+
+  protected def newBlockBody(transactions: immutable.Seq[SignedTransaction], ommers: X): BlockBody
+
+  protected def defaultPrepareHeader(
+    blockNumber: BigInt,
+    parent: Block,
+    beneficiary: Address,
+    blockTimestamp: Long,
+    ommers: Ommers
+  ): BlockHeader =
+      BlockHeader(
+        parentHash = parent.header.hash,
+        ommersHash = ByteString(kec256(ommers.toBytes: Array[Byte])),
+        beneficiary = beneficiary.bytes,
+        stateRoot = ByteString.empty,
+        //we are not able to calculate transactionsRoot here because we do not know if they will fail
+        transactionsRoot = ByteString.empty,
+        receiptsRoot = ByteString.empty,
+        logsBloom = ByteString.empty,
+        difficulty = difficulty.calculateDifficulty(blockNumber, blockTimestamp, parent.header),
```

Comment:
I don't think difficulty makes any sense outside of PoW context. Definitely not for now, but maybe we should follow up on this and avoid this calculation for RAFT?